### PR TITLE
Add a pytest rules

### PR DESCRIPTION
**Pull request by AutomatedTester**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Does not include precompiled binaries, eg. `.par` files. See [CONTRIBUTING.md](https://github.com/bazelbuild/rules_python/blob/master/CONTRIBUTING.md) for info
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by thundergolfer — COMMENTED** [review]
Thanks for kicking this off. Will be taking it for a spin in https://github.com/thundergolfer-playground/rules_python-pr-464

**Comment by joshua-cannon-techlabs** [comment]
Firstly, I would absolutely love to see `pytest` support in the stock `bazel` rules.

However, one thing that worries me is capturing the invisible dependency between test files and `conftest.py` files. Pytest-the-framework will load those to find customizations and fixtures which tests use without a paper-trail. (See [pytest's doc](https://docs.pytest.org/en/latest/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session)).

So if the user is specifying the test, _someone_ (is it them or the framework?) should be declaring those dependencies (`conftest.py` all the way up). I would argue the framework should, as it's _painfully_ easy to get wrong from the user's perspective.

The same goes for the test suit rule.

File: python/pytest.bzl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    return repr(paths)
+
+def _pytest_runner_impl(ctx):
+    if len(ctx.attr.srcs) == 0:
+        fail("No test files specified.")
+
+    expanded_args = [ctx.expand_location(arg, ctx.attr.data) for arg in ctx.attr.args]
+
+    runner = ctx.actions.declare_file(ctx.attr.name)
+    ctx.actions.write(
```
Comment (hrfuller) [inline python/pytest.bzl:13]: This should be a template substitution action.

File: python/pytest.bzl [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        args = args,
+        data = data,
+        python_version = python_version,
+    )
+
+    py_test(
+        name = name,
+        python_version = python_version,
+        srcs = srcs + [runner_target],
+        deps = deps,
```
[the remark this answers is not in the document]
Comment (hrfuller) [reply python/pytest.bzl:76]: Only issue I can think of involves which version of pytest to use. Presumably users need to have it in their requirements.txt, or we inject a default version if not. We should definitely add `requirement("pytest")` to deps if it doesn't exist.

Other related issues include user provided plugin version eg. pytest_coverage which need to resolve with pytest, and how / when to add those deps to the py_pytest_test rule.

**Review by hrfuller — COMMENTED** [review]
I'd love to revive this PR. As a first pass we can't and shouldn't aim to support all the configurability of pytest, but we should let users passthrough pytest args to the stub script that invokes it.

Other things to consider are: 
* How to treat configuration files we find in the test sandbox, easiest thing at first would be to ignore custom configs.
* we should amend the `srcs` of the underlying native.py_test to include a glob for 
**/conftest.py as someone mentioned in a comment.

**Comment by aignas** [comment]
When someone comes back to this, consider adding extra arguments to change how pytest is changing the import path. I did some investigation as to why pytest tests where failing when I was using protobuf rules here: https://github.com/aignas/bazel_pytest_proto

**Comment by betaboon** [comment]
@AutomatedTester are you planning/willing to pick this up again, or would you prefer someone else taking it over?

@everyone else in here: what would be a minimal setup of changes required here to get this merged?

i see the following todo-list:
- address (all) the review-comments
- clear up with we want to implicitly supply a `pytest` and if so how

i would argue we shouldn't aim for adressing all eventualities at first, otherwise this would never get in in any form.

just for clarification: i would be willing to take this over under the condition that there is a willingness for constructive cooperation and eventually compromise.

**Comment by AutomatedTester** [comment]
> @AutomatedTester are you planning/willing to pick this up again, or would you prefer someone else taking it over?

Happy for someone to take this over, I haven't had time to do it.

**Comment by thundergolfer** [comment]
Heads up that https://github.com/bazelbuild/rules_python/pull/723 may supersede this.

**Comment by alexeagle** [comment]
Note, https://docs.aspect.build/rules/aspect_rules_py/docs/rules#py_pytest_main provides missing glue for pytest, thanks to @mattem and @f0rmiga

**Pull request closed without merging**